### PR TITLE
Make sure temporary tables always get cleaned up, even on faillure

### LINKED ISSUE #22273: M2 catalogrule_product__temp Tables
### Preconditions (*)

1. Magento 2.3.1
2. MySQL 5.7 / PHP 7.2

### Steps to reproduce (*)

1. Look at database in phpmyadmin

### Expected result (*)

1. These tables should delete after cart/catalog price rule is disabled/expired/inactive.

### Actual result (*)
1. These tables are not removed after rules are over. Database stays inflated but thankfully doesn't use the rule prices/actions.
Capture

### PR DESCRIPTION
**Pull request by quisse**
### Description (*)
Make sure temporary tables always get cleaned up, even on faillure.

### Fixed Issues (if relevant)

1. Fixes magento/magento2#22273

### Manual testing scenarios (*)

1. ...
2. ...

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Review by kandy — CHANGES_REQUESTED** [review]
I believe there is a drawback to your fix.
After your changes, the index table will be cleaned up if any exception happens 

Better to modify IndexerTableSwapper class to:
 -  Create a temporary table instead of real  in https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/CatalogRule/Model/Indexer/IndexerTableSwapper.php#L50
 - or/and cleanup temporary table in class destructor

File: app/code/Magento/CatalogRule/Model/Indexer/IndexerTableSwapper.php [hunk 1]
```diff
@@ -49,7 +49,7 @@ private function createTemporaryTable(string $originalTableName): string
 
         $this->resourceConnection->getConnection()->query(
             sprintf(
-                'create table %s like %s',
+                'create temporary table %s like %s',
                 $temporaryTableName,
```
Comment (kandy) [inline app/code/Magento/CatalogRule/Model/Indexer/IndexerTableSwapper.php]: Can I ask you slightly improve this 
```suggestion
       $this->resourceConnection->getConnection()->createTemporaryTableLike(
            $temporaryTableName,
            $this->resourceConnection->getTableName($originalTableName)
        );
```

**Review by kandy — CHANGES_REQUESTED** [review]

**Comment by quisse** [comment]
The "Database Compare" (for example) build task failed. All reports available show success. Am I missing something? See:

- https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/27012/c6949b214931a7c9a68034e3765e4038/DatabaseCompare/b2b-skip-db-compare-report.html
- https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/27012/c6949b214931a7c9a68034e3765e4038/DatabaseCompare/ce-db-compare-report.html
- https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/27012/c6949b214931a7c9a68034e3765e4038/DatabaseCompare/ce-skip-db-compare-report.html
- https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/27012/c6949b214931a7c9a68034e3765e4038/DatabaseCompare/ee-db-compare-report.html
- https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/27012/c6949b214931a7c9a68034e3765e4038/DatabaseCompare/ee-skip-db-compare-report.html

**Comment by hostep** [comment]
What's the status here? Just saw an error:
```
Table 'catalogrule_product__temp8f30230c' doesn't exist, query was: SHOW CREATE TABLE `catalogrule_product__temp8f30230c
```
happening on a shop during `bin/magento setup:upgrade` (shop was upgraded to Magento 2.3.5 yesterday).

Will this PR fix this problem?

**Comment by quisse** [comment]
@kandy please, any response would be appreciated.

**Comment by santibm** [comment]
hi, what's the status of this!? it's happening now on two customer sites heavily using catalog rules, @quisse does your fix work?!

**Comment by engcom-Charlie** [comment]
Hi @quisse, could you please look through failed tests?

**Comment by quisse** [comment]
@engcom-Charlie Please read previous comment.

**Review by kandy — DISMISSED** [review]

**Review by engcom-Alfa — CHANGES_REQUESTED** [review]
Hi @quisse   .
     Due to [Magento Definition of Done](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing_dod.html) all code must be covered by tests. Please cover your fix by automated test.
Thanks!

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by MarienLbc** [comment]
Hello!
As I am facing this issue in my project, I tried this patch on my dev environment and it looks some tables are dropped after catalogrule_rule index execution. FYI tables are : 
- catalogrule_group_website
- catalogrule_product
- ​catalogrule_product_price
I'm taking a look at it :eyeglasses:

**Comment by quisse** [comment]
- Database Compare fails -> everything has success
- Functional Tests fails -> doesn't seem to be impacted by my change
- Integration Tests fails -> only warnings?
- Sample Data Tests fails -> "No build reports available"
- WebAPI Tests fails -> everything has success

Do you guys know how bad the [DX](https://medium.com/swlh/what-is-dx-developer-experience-401a0e44a9d9) is when trying to contribute to this OS project?

**Comment by engcom-Charlie** [comment]
Hi @quisse, can you please merge mainline into your branch and re-run tests?
Thank you.

**Comment by quisse** [comment]
@magento run all tests

**Comment by quisse** [comment]
@engcom-Charlie Done, but still same test results.

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by quisse** [comment]
Same result, over and over...
This makes contributing quite unpleasant imho.

**Comment by quisse** [comment]
image

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@sidolov could you provide why some builds are failing without reports available?

**Comment by sidolov** [comment]
```
....Error in fixture: "\/var\/www\/html\/dev\/tests\/integration\/testsuite\/Magento\/CatalogRule\/_files\/rule_apply_as_percentage_of_original_not_logged_user.php".
 Catalog rule indexing failed. See details in exception log.
 #0 /var/www/html/dev/tests/integration/testsuite/Magento/CatalogRule/_files/rule_apply_as_percentage_of_original_not_logged_user.php(43): Magento\CatalogRule\Model\Indexer\IndexBuilder->reindexFull()
#1 /var/www/html/dev/tests/integration/framework/Magento/TestFramework/Annotation/AbstractDataFixture.php(87): require('/var/www/html/d...')
#2 /var/www/html/dev/tests/integration/framework/Magento/TestFramework/Annotation/AbstractDataFixture.php(121): Magento\TestFramework\Annotation\AbstractDataFixture->_applyOneFixture()
#3 /var/www/html/dev/tests/integration/framework/Magento/TestFramework/Annotation/DataFixture.php(35): Magento\TestFramework\Annotation\AbstractDataFixture->_applyFixtures()
#4 [internal function]: Magento\TestFramework\Annotation\DataFixture->startTestTransactionRequest()
#5 /var/www/html/dev/tests/integration/framework/Magento/TestFramework/EventManager.php(49): call_user_func_array()
#6 /var/www/html/dev/tests/integration/framework/Magento/TestFramework/Event/Transaction.php(76): Magento\TestFramework\EventManager->fireEvent()
#7 /var/www/html/dev/tests/integration/framework/Magento/TestFramework/Event/Transaction.php(46): Magento\TestFramework\Event\Transaction->_processTransactionRequests()
#8 [internal function]: Magento\TestFramework\Event\Transaction->startTest()
#9 /var/www/html/dev/tests/integration/framework/Magento/TestFramework/EventManager.php(49): call_user_func_array()
#10 /var/www/html/dev/tests/integration/framework/Magento/TestFramework/Event/PhpUnit.php(127): Magento\TestFramework\EventManager->fireEvent()
#11 /var/www/html/vendor/phpunit/phpunit/src/Framework/TestResult.php(407): Magento\TestFramework\Event\PhpUnit->startTest()
#12 /var/www/html/vendor/phpunit/phpunit/src/Framework/TestResult.php(624): PHPUnit\Framework\TestResult->startTest()
#13 /var/www/html/vendor/phpunit/phpunit/src/Framework/TestCase.php(772): PHPUnit\Framework\TestResult->run()
#14 /var/www/html/vendor/phpunit/phpunit/src/Framework/TestSuite.php(639): PHPUnit\Framework\TestCase->run()
#15 /var/www/html/vendor/phpunit/phpunit/src/Framework/TestSuite.php(639): PHPUnit\Framework\TestSuite->run()
#16 /var/www/html/vendor/phpunit/phpunit/src/Framework/TestSuite.php(639): PHPUnit\Framework\TestSuite->run()
#17 /var/www/html/vendor/phpunit/phpunit/src/Framework/TestSuite.php(639): PHPUnit\Framework\TestSuite->run()
#18 /var/www/html/vendor/phpunit/phpunit/src/TextUI/TestRunner.php(653): PHPUnit\Framework\TestSuite->run()
#19 /var/www/html/vendor/phpunit/phpunit/src/TextUI/Command.php(108): PHPUnit\TextUI\TestRunner->run()
#20 /var/www/html/vendor/phpunit/phpunit/src/TextUI/Command.php(68): PHPUnit\TextUI\Command->run()
#21 /var/www/html/vendor/phpunit/phpunit/phpunit(61): PHPUnit\TextUI\Command::main()
#22 {main}
```

**Comment by ihor-sviziev** [comment]
@sidolov ok, could you also provide info from exception log?

**Comment by sidolov** [comment]
@ihor-sviziev 
```
SQLSTATE[42S02]: Base table or view not found: 1146 Table 'magento2_prod.catalogrule_product__tempc886e889' doesn't exist, query was: RENAME TABLE catalogrule_product TO catalogrule_product34959eff,catalogrule_product__tempc886e889 TO catalogrule_product,catalogrule_product_price TO catalogrule_product_price33f11c26,catalogrule_product_price__temp1611874b TO catalogrule_product_price,catalogrule_group_website TO catalogrule_group_website310d948a,catalogrule_group_website__tempbaba801f TO catalogrule_group_website
```

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @quisse,
Seems like your change broke indexation mechanism, you can see the error that appeared here https://github.com/magento/magento2/pull/27012#issuecomment-696916060

Could you review and fix this issue?

**Comment by engcom-Charlie** [comment]
Hi @quisse, can you please look at https://github.com/magento/magento2/pull/27012#pullrequestreview-493760614? Otherwise, we can't proceed with your PR.
Thank you.

**Comment by quisse** [comment]
It's almost been **half a year** since I worked on this PR. It took **half a year** to get proper feedback. I've been quite patient from my side. I would appreciate it if you guys now are also patient. I do this in my spare time.

**Comment by ihor-sviziev** [comment]
Hi @quisse,
👍 Sure, no rush at all

**Comment by quisse** [comment]
@magento run all tests

**Comment by quisse** [comment]
@magento run all tests

**Comment by quisse** [comment]
Updated the solution to be without temporary tables but with a cleanup afterwards because the "temporary" tables are swapped to non temporary tables.
I'm not sure if the destructor call needs to be tested?

**Comment by quisse** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run Integration Tests, Functional Tests B2B

**Comment by ihor-sviziev** [comment]
@magento run Performance Acceptance Tests

**Comment by ihor-sviziev** [comment]
Seems like test failures not related to changes from this PR. Let’s wait a bit and restart tests again

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B, Integration Tests

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B

**Comment by quisse** [comment]
Feels like a lottery, isn't it? 🙈😅

**Comment by ihor-sviziev** [comment]
There were some fixes in related repositories, so tests should be fixed already

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B

**Comment by quisse** [comment]
@magento run Functional Tests B2B

**Comment by ihor-sviziev** [comment]
@quisse could you merge recent changes from 2.4-develop? Maybe that's causing test failures?

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8367](https://jira.corp.magento.com/browse/ENGCOM-8367) has been created to process this Pull Request

**Comment by ihor-sviziev** [comment]
@engcom-Alfa any updates on testing this PR?

**Review by jordanngamba — APPROVED** [review]

**Comment by stefandoorn** [comment]
This change will still leave temporary tables in case Magento hit a PHP memory limit, which is not that hard in Magento landscape. Maybe the indexer should start by cleaning up old tables, instead of doing it in the destructor.

**Comment by ihor-sviziev** [comment]
@stefandoorn, that's actually another use case. From my perspective - this solution is good enough to be tested and merged. If you have a better solution - please send a Pull Request, we'll review it

**Comment by stefandoorn** [comment]
Yeah, sorry, wasn't meant to discard this PR. It's indeed another use case to keep in mind :-)

**Comment by vzabaznov** [comment]
Hey @quisse, the Performance team took care of your PR. Thank you for your contribution.

**Pull request merged**